### ISSUE TITLE
fit_gpytorch_model error for gpytorch 1.3.1 with pytest

### ISSUE THREAD

**Issue by DavidWalz** [pytorch/botorch#668]
# 🐛 Bug

Since upgrading to gpytorch 1.3.1 we're seeing a weird behavior that many of our unit test on botorch-based components fail, but work when executed directly. I broke it down to the following minimal failing example:
```
import torch
from botorch.fit import fit_gpytorch_model
from botorch.models import SingleTaskGP
from gpytorch.mlls import ExactMarginalLogLikelihood

def test_fit():
    X = torch.tensor([0, 0.25, 0.5, 0.75, 1.0], dtype=torch.float)
    Y = torch.tensor(
        [0.13721396, 1.14897599, 1.49943135, 1.35984674, 0.69536624], dtype=torch.float
    )
    model = SingleTaskGP(X.view(-1, 1), Y.view(-1, 1))
    mll = ExactMarginalLogLikelihood(model.likelihood, model)
    fit_gpytorch_model(mll)

if __name__ == "__main__":
    test_fit()
```

Running the script works, but running it with pytest fails.
The data is random. Basically a lot of the tests involving random data suddenly fail with high probability.
I realize this is very likely a gpytorch issue, but it's not straightforward to factor out the botorch components in the above example.

** Stack trace/error message **
```
pytest .\mnwe.py 
==================================================================== test session starts =====================================================================
platform win32 -- Python 3.8.3, pytest-6.2.1, py-1.9.0, pluggy-0.13.1

plugins: cov-2.10.1
collected 1 item

mnwe.py F                                                                                                                                               [100%]

========================================================================== FAILURES ========================================================================== 
__________________________________________________________________________ test_fit __________________________________________________________________________ 

    def test_fit():
        X = torch.tensor([0, 0.25, 0.5, 0.75, 1.0], dtype=torch.float)
        Y = torch.tensor(
            [0.13721396, 1.14897599, 1.49943135, 1.35984674, 0.69536624], dtype=torch.float
        )
        model = SingleTaskGP(X.view(-1, 1), Y.view(-1, 1))
        mll = ExactMarginalLogLikelihood(model.likelihood, model)
>       fit_gpytorch_model(mll)

mnwe.py:14:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  
..\..\..\miniconda3\lib\site-packages\botorch\fit.py:125: in fit_gpytorch_model
    sample_all_priors(mll.model)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  

model = SingleTaskGP(
  (likelihood): GaussianLikelihood(
    (noise_covar): HomoskedasticNoise(
      (noise_prior): GammaPri...ance_module): Distance()
    )
    (outputscale_prior): GammaPrior()
    (raw_outputscale_constraint): Positive()
  )
)

    def sample_all_priors(model: GPyTorchModel) -> None:
        r"""Sample from hyperparameter priors (in-place).

        Args:
            model: A GPyTorchModel.
        """
>       for _, prior, closure, setting_closure in model.named_priors():
E       ValueError: too many values to unpack (expected 4)

..\..\..\miniconda3\lib\site-packages\botorch\optim\utils.py:37: ValueError
====================================================================== warnings summary ====================================================================== 
mnwe.py::test_fit
  ...\miniconda3\lib\site-packages\botorch\models\utils.py:189: InputDataWarning: Input data is not standardized. Please consider scaling the input to zero mean and unit variance.
    warnings.warn(msg, InputDataWarning)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
================================================================== short test summary info =================================================================== 
FAILED mnwe.py::test_fit - ValueError: too many values to unpack (expected 4)
================================================================ 1 failed, 1 warning in 3.95s ================================================================
```
## System information
python 3.8.3 and python 3.8.5
torch 1.7.1+cpu
gpytorch 1.3.1
botorch 0.3.3
pytest 6.2.1
We're also seeing this in our CI on python 3.8.5.

**Comment by Balandat** [comment]
Thanks for raising this. Taking a look now. FWIW botorch 0.3.3. hasn't been properly tested with the most recent GPyTorch release, so no guarantees :)

**Comment by Balandat** [comment]
Ah ok I see what's going on - The error you're seeing here is due a GPyTorch breaking change to priors that was necessary for properly copying models (incidentally, I made that change): https://github.com/cornellius-gp/gpytorch/pull/1361

In your failures you should see a message like:
> ValueError: As of version 1.4, `param_or_closure` must operate on a module instance. For example: [...]

This all works on botorch master. We plan on releasing a new version with those changes within the next couple of weeks. In the meantime, you have two options:
1. Pin gpytorch to 1.3.0
2. Use botorch master

> The data is random. Basically a lot of the tests involving random data suddenly fail with high probability.

This is something I don't observe even on 0.3.3, all my test failures are related to the changes to the priors. Let me know if this is reproducible.

**Comment by DavidWalz** [comment]
Thanks for the unbeatably quick reply. We're pinning to 1.3.0 for now.

> > The data is random. Basically a lot of the tests involving random data suddenly fail with high probability.

> This is something I don't observe even on 0.3.3, all my test failures are related to the changes to the priors. Let me know if this is reproducible.

I spent an hour boiling down the test to see what makes it break. In the beginning I was using random X and Y and since there was a chance that the test actually ran through, I looped over it a couple of time. Once I fixed the specific values in the test above, it failed reproducibly.

**Comment by Balandat** [comment]
Closing this out since this is addressed on master and will be fixed in the upcoming 0.4.0 release
